lib/streamlit/util.py

```diff
@@ -203,3 +216,37 @@ def extract_key_query_params(
             for item in sublist
         ]
     )
+
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+class TimedCleanupCache(TTLCache, Generic[K, V]):
+    """A TTLCache that asynchronously expires its entries."""
+
+    def __init__(self, *args, **kwargs):
+        self._task: Optional[asyncio.Task[Any]] = None
+        super().__init__(*args, **kwargs)
+
+    def __setitem__(self, key: K, value: V) -> None:
+        # Set an expiration task to run periodically
+        # Can't be created in init because that only runs once and
+        # the event loop might not exist yet.
+        if self._task is None:
+            try:
+                self._task = asyncio.create_task(expire_cache(self))
+            except RuntimeError:
+                # Just continue if the event loop isn't started yet.
+                pass
+        super().__setitem__(key, value)
+
+    def __del__(self):
+        if self._task is not None:
+            self._task.cancel()
+
+
+async def expire_cache(cache: TTLCache) -> None:
+    while True:
+        await asyncio.sleep(30)
+        cache.expire()
```

Comment:
Is `cache.expire()` an in-expensive operation?  We could go more conservative by using longer sleep, but I assume calling `cache.expire()` isn't really expensive, or?

Reply:
I think it's cheap, because it iterates over the entries in order of expiration time and stops when it finds something live, so most of the time it does almost no work